### PR TITLE
grid method for implicit regressor inference

### PR DESCRIPTION
**Pull request by tomsilver**
`python src/main.py --approach nsrt_learning --env touch_point --option_learner implicit_bc --implicit_mlp_regressor_inference_method grid` 50/50 on the first 5 seeds

### CODE REVIEW

File: src/torch_models.py [hunk 1]
```diff
@@ -351,6 +351,11 @@ class ImplicitMLPRegressor(PyTorchRegressor):
     inputs and returning the sample that has the highest probability of
```
Comment (ronuchit) [inline src/torch_models.py:351]: returns*

File: src/torch_models.py [hunk 2]
```diff
@@ -508,6 +517,25 @@ def _predict_sample_once(self, x: Array) -> Array:
         sample_idx = torch.argmax(scores)
         return sample_ys[sample_idx]
 
+    def _predict_grid(self, x: Array) -> Array:
+        assert self._grid_num_ticks_per_dim is not None
+        assert self._grid_num_ticks_per_dim > 0
+        dy = 1.0 / self._grid_num_ticks_per_dim
+        ticks = [np.arange(0.0, 1.0, dy)] * self._y_dim
+        grid = np.meshgrid(*ticks)  # type: ignore
+        candidate_ys = np.transpose(grid).reshape((-1, self._y_dim))
```
Comment (ronuchit) [inline src/torch_models.py:526]: isn't the grid symmetric? what does this transpose do?

File: src/torch_models.py [hunk 3]
```diff
@@ -351,6 +351,11 @@ class ImplicitMLPRegressor(PyTorchRegressor):
     inputs and returning the sample that has the highest probability of
     classifying to 1, under the learned classifier.
 
+    Inference with the "grid" method is similar to "sample_once", except that
```
Comment (ronuchit) [inline src/torch_models.py:354]: move this para down after derivative_free one, for ordering consistency

File: src/torch_models.py [hunk 4]
```diff
@@ -488,6 +495,8 @@ def _predict(self, x: Array) -> Array:
         assert x.shape == (self._x_dim, )
         if self._inference_method == "sample_once":
             return self._predict_sample_once(x)
+        if self._inference_method == "grid":
```
Comment (ronuchit) [inline src/torch_models.py:498]: move this down after derivative_free one, for ordering consistency

**Review by ronuchit — APPROVED** [review]

File: src/torch_models.py (hunk 2 above)
Comment (tomsilver) [reply src/torch_models.py:526]: without the transpose, the shape would be (num_samples, y_dim) instead of (y_dim, num_samples)

**Pull request merged**